File: mobile/android/app/src/main/java/com/example/seniordesignmobileapp/modeling/NodePoseMath.kt
package com.example.seniordesignmobileapp.modeling

import com.example.seniordesignmobileapp.model.ImuSample
import kotlin.math.atan2
import kotlin.math.pow
import kotlin.math.sqrt

data class NodePoseEstimate(
    val pitchDeg: Float,
    val rollDeg: Float,
    val gravityVector: NormalizedVector3,
)

data class NormalizedVector3(
    val x: Float,
    val y: Float,
    val z: Float,
)

object NodePoseMath {
    fun fromSample(
        sample: ImuSample,
    ): NodePoseEstimate? {
        val ax = sample.ax.toDouble()
        val ay = sample.ay.toDouble()
        val az = sample.az.toDouble()
        val magnitude = sqrt(ax.pow(2) + ay.pow(2) + az.pow(2))
        if (magnitude == 0.0) {
            return null
        }

        val gravityVector = NormalizedVector3(
            x = (ax / magnitude).toFloat(),
            y = (ay / magnitude).toFloat(),
            z = (az / magnitude).toFloat(),
        )
        val pitchDeg = Math.toDegrees(
            atan2(
                gravityVector.x.toDouble(),
                sqrt(gravityVector.y.toDouble().pow(2) + gravityVector.z.toDouble().pow(2)),
            )
        ).toFloat()
        val rollDeg = Math.toDegrees(
            atan2(
                gravityVector.y.toDouble(),
                gravityVector.z.toDouble(),
            )
        ).toFloat()

        return NodePoseEstimate(
            pitchDeg = pitchDeg,
            rollDeg = rollDeg,
            gravityVector = gravityVector,
        )
    }
}

File: mobile/ios/FH07Mobile/Modeling/NodePoseMath.swift
import Foundation

struct NodePoseEstimate: Equatable {
    let pitchDeg: Float
    let rollDeg: Float
    let gravityVector: NormalizedVector3
}

struct NormalizedVector3: Equatable {
    let x: Float
    let y: Float
    let z: Float
}

enum NodePoseMath {
    static func fromSample(_ sample: ImuSample) -> NodePoseEstimate? {
        let ax = Double(sample.ax)
        let ay = Double(sample.ay)
        let az = Double(sample.az)
        let magnitude = (ax * ax + ay * ay + az * az).squareRoot()
        if magnitude == 0 { return nil }

        let gravityVector = NormalizedVector3(
            x: Float(ax / magnitude),
            y: Float(ay / magnitude),
            z: Float(az / magnitude)
        )
        let gx = Double(gravityVector.x)
        let gy = Double(gravityVector.y)
        let gz = Double(gravityVector.z)
        let pitchDeg = Float(atan2(gx, (gy * gy + gz * gz).squareRoot()) * 180.0 / .pi)
        let rollDeg = Float(atan2(gy, gz) * 180.0 / .pi)

        return NodePoseEstimate(pitchDeg: pitchDeg, rollDeg: rollDeg, gravityVector: gravityVector)
    }
}
